### PR TITLE
CORELLI tube calibration: switch acceptance criterion

### LINKED ISSUE #29766: CORELLI tube calibration: switch acceptance criterion
See https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/122 for a description

### PR DESCRIPTION
**Pull request by jmborr**
Signed-off-by: Jose Borreguero <borreguero@gmail.com>

**Description of work.**
- [x] criterion function using the vertical (Y-coordinate) fitted positions of the wire shadows
- [x] switch criteria functions (from pixel to vertical positions) in all dependent functions

**To test:**
First make sure that the `scripts/` directory located under the Mantid source repository is added to the list of scripts directories. Then, in the script Editor of the workbench, run the following script (should take not more than 5 minutes):
```
from corelli.calibration.bank import calibrate_banks
from corelli.calibration.utils import apply_calibration, load_banks
counts_file = '/SNS/CORELLI/shared/tmp/CORELLI_124023_counts.nxs'
load_banks(counts_file, bank_selection='10-19', output_workspace='counts')
calibrate_banks('counts', bank_selection='10-19')
apply_calibration('counts', 'calib10', output_workspace='counts_10')
apply_calibration('counts', 'calib14', output_workspace='counts_14')
```
Open the instrument view for workspaces `counts` and `counts_10`, then look at `bank10` and verify that all tubes in this bank have been calibrated for `counts_10`. Before switching criterion, `tube13` [was not successfully calibrated](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/121).

Capture

Open the instrument view for workspaces `counts` and `counts_10`, then look at `bank10` and verify that all tubes in this bank have been calibrated for `counts_10`

Fixes #29766

This does not require release notes because it's a protected change not exposed to the User

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/corelli/calibration/bank.py [hunk 1]
```diff
@@ -180,10 +186,83 @@ def collect_bank_fit_results(output_workspace: str,
     return workspace
 
 
-def criterium_peak_pixel_position(peak_table: InputTable,
+def criterion_peak_vertical_position(peak_table: InputTable,
+                                     summary: Optional[str] = None,
+                                     zscore_threshold: float = 2.5,
+                                     deviation_threshold: float = 0.0035) -> np.ndarray:
```
Comment (williamfgc) [inline scripts/corelli/calibration/bank.py:192]: I'd just add a note on hardcoded defaults

File: scripts/corelli/calibration/bank.py [hunk 2]
```diff
@@ -338,16 +418,21 @@ def calibrate_bank(workspace: WorkspaceTypes,
     table of calibrated detector IDs and a table of non-calibrated detector IDs
 
     This function produces the following temporary workspaces: 'PeakTable' (TableWorkspace),
-    'parameters_table' (WorkspaceGroup), and 'acceptance' (Workspace2D). Thus, any extant workspace having
-    any of these names will be overwritten
+    'PeakTable' (TableWorkspace), 'ParametersTable' (WorkspaceGroup), and 'acceptance' (Workspace2D).
```
Comment (williamfgc) [inline scripts/corelli/calibration/bank.py:421]: I believe we switched from acceptance to "fit"

File: scripts/corelli/calibration/bank.py [hunk 3]
```diff
@@ -180,10 +186,83 @@ def collect_bank_fit_results(output_workspace: str,
...
+    :param zscore_threshold: maximum Z-score for the vertical positions of a tube.
+    :param deviation_threshold: maximum deviation (in meters) for the vertical positions of the wire shadows.
+    :return: array of booleans, one per tube. `True` is the tube passes the acceptance criterion, `False` otherwise.
+    """
+    table = mtd[str(peak_table)]  # handle to the peak table
+    peak_count = table.columnCount() - 1  # the first column contains the names of the tubes
+    # `positions_average` stores the vertical position for each wire shadow, averaged for all tubes
+    positions_average = [np.mean(table.column(column_number)) for column_number in range(1, 1 + peak_count)]
+
+    deviations = list()  # a measure of how much the vertical positions deviate from the mean positions
```
Comment (williamfgc) [inline scripts/corelli/calibration/bank.py:214]: just bikeshedding,  `deviation:List[np.array] = []` is more clear

File: scripts/corelli/calibration/bank.py [hunk 4]
```diff
@@ -180,10 +186,83 @@ def collect_bank_fit_results(output_workspace: str,
...
+        workspace.replaceAxis(1, axis)
+
+    return criterion_pass
+
+
+def criterion_peak_pixel_position(peak_table: InputTable,
                                   summary: Optional[str] = None,
                                   zscore_threshold: float = 2.5,
-                                  deviation_threshold: float = 3.0) -> np.ndarray:
+                                  deviation_threshold: float = 0.0035) -> np.ndarray:
```
Comment (williamfgc) [inline scripts/corelli/calibration/bank.py:265]: I'd add a note on the hardcoded value when documenting parameters

File: scripts/corelli/calibration/utils.py [hunk 5]
```diff
@@ -7,12 +7,14 @@
 
 from os import path
 import numpy as np
+from numpy.polynomial.polynomial import polyval
 from typing import List, Optional, Union
```
Comment (williamfgc) [inline scripts/corelli/calibration/utils.py:11]: Not sure if `List` is used as a type

File: scripts/test/corelli/calibration/test_utils.py [hunk 6]
```diff
@@ -104,24 +130,73 @@ def test_load_banks(self):
         self.assertAlmostEqual(workspace.readY(42)[0], 13297.0)
         DeleteWorkspaces(['jambalaya'])
 
-    def test_calculate_tube_calibration(self) -> None:
+    def test_peak_y_table(self) -> None:
+        # Mock PeakTable with two tubes and three peaks. Simple, integer values
+        def peak_pixels_table(table_name, peak_count, tube_names=None, pixel_positions=None):
```
Comment (williamfgc) [inline scripts/test/corelli/calibration/test_utils.py:135]: add return type `->TableWorkspace` to function signature

File: scripts/test/corelli/calibration/test_utils.py [hunk 7]
```diff
@@ -104,24 +130,73 @@ def test_load_banks(self):
...
+                for tube_index in range(len(tube_names)):
+                    # tube_names is a list of str values; pixel_positions is a list of lists of float values
+                    table.addRow([tube_names[tube_index]] + pixel_positions[tube_index])
+            return table
+
+        # Create a peak table with only one tube
+        peak_table = peak_pixels_table('PeakTable', 3, ['tube1'], [[0, 1, 2]])
+
+        # Mock ParametersTableGroup with one parameter table. Simple parabola
+        def parameters_optimized_table(table_name, values=None, errors=None):
```
Comment (williamfgc) [inline scripts/test/corelli/calibration/test_utils.py:151]: `->TableWorkspace` as return type

File: scripts/test/corelli/calibration/test_utils.py [hunk 8]
```diff
@@ -10,14 +10,34 @@
...
+    @classmethod
+    def setUpClass(cls) -> None:
+        r"""
+        Load the tests cases for calibrate_bank, consisting of data for only one bank
+        CORELLI_124023_bank10, tube 13 has shadows at pixel numbers quite different from the rest
+        """
+        config.appendDataSearchSubDir('CORELLI/calibration')
+        for directory in config.getDataSearchDirs():
+            if 'UnitTest' in directory:
+                data_dir = path.join(directory, 'CORELLI', 'calibration')
```
Comment (williamfgc) [inline scripts/test/corelli/calibration/test_utils.py:31]: Just bikeshedding, I think `pathlib` is preferred these days on Python 3.

File: scripts/corelli/calibration/utils.py [hunk 9]
```diff
@@ -91,13 +93,67 @@ def load_banks(filename: str, bank_selection: str, output_workspace: str) -> Wor
...
+
+
+def calibrate_tube(workspace: WorkspaceTypes,
+                   tube_name: str,
+                   output_peak_table: str = 'PeakTable',
+                   output_parameters_table: str = 'ParametersTable',
+                   output_peak_y_table: str = 'PeakYTable',
+                   shadow_height: float = 1000,
+                   shadow_width: float = 4,
+                   fit_domain: float = 7) -> TableWorkspace:
```
Comment (williamfgc) [inline scripts/corelli/calibration/utils.py:144]: Just add comment for hardcoded values, unless they are obvious.

**Review by williamfgc — CHANGES_REQUESTED** [review]
Overall, is good. I also tested locally with MantidWorkbench and works as expected.

File: scripts/corelli/calibration/bank.py (hunk 1 above)
Comment (jmborr) [reply scripts/corelli/calibration/bank.py:192]: I described in the docstring that `0.0035` corresponds to the height of three CORELLI pixels

File: scripts/corelli/calibration/bank.py (hunk 4 above)
Comment (jmborr) [reply scripts/corelli/calibration/bank.py:265]: actually this is wrong. "3" is correct for function `criterion_peak_pixel_position`
I describe now in the docstring that the default of `3` corresponds to the height of three CORELLI pixels

File: scripts/corelli/calibration/bank.py (hunk 2 above)
Comment (jmborr) [reply scripts/corelli/calibration/bank.py:421]: In this context, the `acceptance` table is appropriate. The `fit` table is a merge of the `acceptance` table and the `parameters` table.
The `acceptance` table contains information about the merit of each tube-calibration. The `parameters` table lists some optimized fitting parameters for each tube-calibration. Feng wanted to save both pieces of info so they all go into the `fit` table.

File: scripts/corelli/calibration/utils.py (hunk 5 above)
Comment (jmborr) [reply scripts/corelli/calibration/utils.py:11]: it's not a built-in type, if that's what you mean, but `List` it's a declaration of built-in type `list`

File: scripts/corelli/calibration/bank.py (hunk 2 above)
Comment (williamfgc) [reply scripts/corelli/calibration/bank.py:421]: Got it, thanks!

File: scripts/corelli/calibration/utils.py (hunk 5 above)
Comment (williamfgc) [reply scripts/corelli/calibration/utils.py:11]: yup, I mean type a variable type `list::List[int]`. I'm just used to seeing types everywhere, not a big deal for this.

File: scripts/corelli/calibration/utils.py (hunk 9 above)
Comment (jmborr) [reply scripts/corelli/calibration/utils.py:144]: I introduce an explanation of the hardcoded values for `shadow_height`, `shadow_width`, and `fit_domain` in the docstring of the function

File: scripts/test/corelli/calibration/test_utils.py (hunk 8 above)
Comment (jmborr) [reply scripts/test/corelli/calibration/test_utils.py:31]: I guess so, but `os.path` does the job and I'm too lazy :smile:

File: scripts/test/corelli/calibration/test_utils.py (hunk 6 above)
Comment (jmborr) [reply scripts/test/corelli/calibration/test_utils.py:135]: I have included the annotation, although as a rule, I avoid type annotations in tests :sleepy:

File: scripts/test/corelli/calibration/test_utils.py (hunk 8 above)
Comment (williamfgc) [reply scripts/test/corelli/calibration/test_utils.py:32]: Yeah, going from Python 2 to Python 3 was painful enough. No worries, it's just me bikeshedding.

**Review by williamfgc — DISMISSED** [review]

**Review by williamfgc — APPROVED** [review]

**Pull request merged**